### PR TITLE
refactor: remove deprecated lifecycle methods in Table

### PR DESCRIPTION
**Pull request by afc163**
### 🤔 This is a ...

- [ ] New feature
- [ ] Bug fix
- [ ] Site / document update
- [ ] Component style update
- [ ] TypeScript definition update
- [X] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Other (about what?)

### 🔗 Related issue link

https://github.com/ant-design/ant-design/issues/9792

#18250

### 💡 Background and solution

Updated to `react` v16.9 and it started to log warnings about the unsafe lifecycle methods.

<details>
<summary>📸 Screenshot</summary>

image

</details>

I ran `npx react-codemod rename-unsafe-lifecycles`, following the React [blog post](https://reactjs.org/blog/2019/08/08/react-v16.9.0.html#renaming-unsafe-lifecycle-methods).

### 📝 Changelog

No more warnings about `componentWillReceiveProps` from `Table` and `SelectionCheckboxAll` components. 🙅‍♂ 

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |     -      |
| 🇨🇳 Chinese |     -      |

### ☑️ Self Check before Merge

- [X] Doc is updated/provided or not needed
- [X] Demo is updated/provided or not needed
- [X] TypeScript definition is updated/provided or not needed
- [X] Changelog is provided or not needed

### CODE REVIEW

File: components/table/Table.tsx [hunk 1]
```diff
@@ -87,7 +90,65 @@ const defaultPagination = {
...
+const createComponents = (
+  components: TableComponents = {},
+  prevComponents?: TableComponents,
+  isCalledFromConstructor?: boolean,
+) => {
+  const bodyRow = components && components.body && components.body.row;
+  const prevBodyRow = prevComponents && prevComponents.body && prevComponents.body.row;
+  if (isCalledFromConstructor) {
+    // 'tableRow' needs to act like a class property
+    tableRow = undefined;
```
Comment (orzyyyy) [inline components/table/Table.tsx]: I don't quite understand why adding this, what I think `tableRow` is `undefined` in constructor already. Any other where the `tableRow` had been used?

Comment (afc163) [reply components/table/Table.tsx]: @kristof0425

File: components/table/Table.tsx [hunk 2]
```diff
@@ -148,89 +270,38 @@ export default class Table<T> extends React.Component<TableProps<T>, TableState<
...
-        selectionDirty: false,
-      });
-    }
-
-    // https://github.com/ant-design/ant-design/issues/10133
-    this.CheckboxPropsCache = {};
-
-    if (this.getSortOrderColumns(this.columns).length > 0) {
-      const sortState = this.getSortStateFromColumns(this.columns);
+  componentDidUpdate() {
```
Comment (shaodahong) [inline components/table/Table.tsx]: componentDidUpdate lifecycle maybe not need. getDerivedStateFromProps do same thing

File: components/table/Table.tsx [hunk 3]
```diff
@@ -124,22 +185,83 @@ export default class Table<T> extends React.Component<TableProps<T>, TableState<
...
 
-  columns: ColumnProps<T>[];
+    if ('pagination' in nextProps || 'pagination' in prevProps) {
+      const newPagination = {
+        ...defaultPagination,
+        ...prevState.pagination,
+        ...nextProps.pagination,
+      };
+      newPagination.current = newPagination.current || 1;
+      newPagination.pageSize = newPagination.pageSize || 10;
```
Comment (yoyo837) [inline components/table/Table.tsx]: pageSizeOptions[0] ?

File: components/table/Table.tsx (hunk 2 above)
Comment (orzyyyy) [reply components/table/Table.tsx]: Maybe `getSnapshotBeforeUpdate` is more suitable

File: components/table/Table.tsx (hunk 1 above)
Comment (afc163) [reply components/table/Table.tsx]: https://github.com/ant-design/ant-design/pull/18953/commits/a5ad47b4870af0d510798ba0193fdcf9eb102265

**Comment by orzyyyy** [comment]
看上去只是重构，为啥不发到 `master` 去？

**Comment by afc163** [comment]
@orzyyyy

比较大的重构，稳妥起见放到 minor 上。

**Review by orzyyyy — DISMISSED** [review]

**Review by zombieJ — DISMISSED** [review]

**Comment by afc163** [comment]
谁在帮忙审核一下，给一个 approve。

**Review by orzyyyy — APPROVED** [review]

**Comment by patrickomeara** [comment]
@kristof0425 This is a great PR, well written. I've just updated from `v3.23.6` but wherever I am using `<Table />` with `TableProps` now wants `store`, `checkboxPropsCache`, and `setCheckboxPropsCache`. 

Was there a reason these weren't made optional?

**Comment by javierfz1980** [comment]
I have the same issue reported by @patrickomeara <Table /> now requires store, checkboxPropsCache, setCheckboxPropsCache as mandatory...

**Comment by suraneti** [comment]
I got warning componentWillReceiveProps from using the `notification` component on the latest antd.

**Pull request merged**